compiler/src/iree/compiler/Dialect/Stream/IR/StreamOpFolders.cpp

```diff
@@ -1796,7 +1796,8 @@ void AsyncFillOp::getCanonicalizationPatterns(RewritePatternSet &results,
 //===----------------------------------------------------------------------===//
 
 OpFoldResult AsyncUpdateOp::fold(FoldAdaptor operands) {
-  if (getUpdateSize() == getTargetSize()) {
+  if (getUpdateSize() == getTargetSize() &&
+      getUpdate().getType() == getType()) {
```

Comment:
Why do we need the change?

Reply:
It's to fix a bug exposed by a new assertion in the constant folding code -- we can only replace when types match.